core/src/main/java/org/apache/brooklyn/core/mgmt/classloading/ClassLoaderFromStackOfBrooklynClassLoadingContext.java

```diff
@@ -0,0 +1,137 @@
+ * and the worker wants to have a simple view of the loader to use at any point in time.
+ * For example XStream keeps a class-loader in effect but when deserializing things 
+ * some of those things may define bundles to use. */
+public class ClassLoaderFromStackOfBrooklynClassLoadingContext extends ClassLoader {
+    
+    private final Stack<BrooklynClassLoadingContext> contexts = new Stack<BrooklynClassLoadingContext>();
+    private final Stack<ClassLoader> cls = new Stack<ClassLoader>();
+    private final AtomicReference<Thread> lockOwner = new AtomicReference<Thread>();
+    private ManagementContext mgmt;
+    private ClassLoader currentClassLoader;
+    private AtomicReference<ClassLoaderUtils> currentLoader = new AtomicReference<>();
+    private int lockCount;
+    
+    public ClassLoaderFromStackOfBrooklynClassLoadingContext(ClassLoader classLoader) {
+        setCurrentClassLoader(classLoader);
+    }
+    
+    public void setManagementContext(ManagementContext mgmt) {
+        this.mgmt = checkNotNull(mgmt, "mgmt");
+        currentLoader.set(new ClassLoaderUtils(currentClassLoader, mgmt));
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        return currentLoader.get().loadClass(name);
+    }
+    
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        return findClass(name);
+    }
+
+    /** Must be accompanied by a corresponding {@link #popClassLoadingContext()} when finished. */
+    @SuppressWarnings("deprecation")
+    public void pushClassLoadingContext(BrooklynClassLoadingContext clcNew) {
+        acquireLock();
+        BrooklynClassLoadingContext oldClc;
+        if (!contexts.isEmpty()) {
+            oldClc = contexts.peek();
+        } else {
+            // TODO xml serializers using this should take a BCLC instead of a CL
+            oldClc = JavaBrooklynClassLoadingContext.create(mgmt, getCurrentClassLoader());
+        }
+        BrooklynClassLoadingContextSequential clcMerged = new BrooklynClassLoadingContextSequential(mgmt, oldClc, clcNew);
+        ClassLoader newCL = ClassLoaderFromBrooklynClassLoadingContext.of(clcMerged);
+        contexts.push(clcMerged);
+        cls.push(getCurrentClassLoader());
+        setCurrentClassLoader(newCL);
+    }
+
+    public void popClassLoadingContext() {
+        synchronized (lockOwner) {
+            releaseXstreamLock();
+            setCurrentClassLoader(cls.pop());
+            contexts.pop();
+        }
+    }
+    
+    private ClassLoader getCurrentClassLoader() {
+        return currentClassLoader;
+    }
+    
+    private void setCurrentClassLoader(ClassLoader classLoader) {
+        currentClassLoader = checkNotNull(classLoader);
+        currentLoader.set(new ClassLoaderUtils(currentClassLoader, mgmt));
+    }
+    
+    protected void acquireLock() {
+        synchronized (lockOwner) {
```

Comment:
TL;DR: ignore this comment in this PR!

The threading and acquire/release feels confusing here. It would be good to document the assumptions/behaviour, I think. For example, calling acquire from a different thread has pushed onto the stack, until that other thread has popped everything. Do we really need that? Is the behaviour we want more like thread-local storage, with a different stack for each thread - or do we deliberately want other threads to see a shared `getCurrentClassLoader()` but not be allowed to push onto the stack?

Actually, I see you mostly just moved this code from `XmlMementoSerializer.OsgiClassLoader` so it's fine to leave as-is rather than trying to document or change it more here!

(but it is strange to have "xstream" in the method names, now that it's moved to package `org.apache.brooklyn.core.mgmt.classloading`).